python/private/pypi/whl_library.bzl

```diff
@@ -255,6 +259,17 @@ def _whl_library_impl(rctx):
             logger = logger,
         )
 
+        # If the requirement was a local directory, then we need to watch its content
+        # to recreate the repository when it is modified.
+        # Assume that such packages are imported using a single line requirement
+        # of the form [<name> @] file://<path>
+        # We might have to perform some substitutions in the string before searching.
+        subst_req = envsubst(rctx.attr.requirement, environment.keys(), lambda x, dft: environment[x])
+        if subst_req.startswith("file://"):
+            _, path = subst_req.split("file://", 1)
+            logger.info(lambda: "watching tree {} for wheel library {}".format(path, rctx.name))
+            rctx.watch_tree(path)
```

Comment:
I am not sure if this is the correct behaviour - will we start watching more than just the intended python files? What if the `path` is not a dir? The `watch_tree` will fail.

Reply:
Yes if this is a directory then the correct behaviour should be to fall back to what #2345, or to use `watch` instead of `watch_tree`. Yes, it could end up watching more than the intended python files but as far as I know, `pip` simply copies all of the files as well so I think this is the correct thing to do in that sense. This is definitely a trade-off in that using a local package means giving up on precise dependency tracking of the files within that package.